packages/flutter/lib/src/widgets/gesture_detector.dart

```diff
@@ -924,90 +932,37 @@ class RawGestureDetectorState extends State<RawGestureDetector> {
     return widget.child == null ? HitTestBehavior.translucent : HitTestBehavior.deferToChild;
   }
 
-  void _handleSemanticsTap() {
-    final TapGestureRecognizer recognizer = _recognizers[TapGestureRecognizer];
-    assert(recognizer != null);
-    if (recognizer.onTapDown != null)
-      recognizer.onTapDown(TapDownDetails());
-    if (recognizer.onTapUp != null)
-      recognizer.onTapUp(TapUpDetails());
-    if (recognizer.onTap != null)
-      recognizer.onTap();
+  void _assignSemantics(RenderSemanticsGestureHandler renderObject) {
+    if (_semanticsMapping == null)
+      return;
+    renderObject
+      ..onTap = _semanticsMapping.getTapHandler(_findRecognizer) == null
+          ? null : _handleSemanticTap
+      ..onLongPress = _semanticsMapping.getLongPressHandler(_findRecognizer) == null
+          ? null : _handleSemanticLongPress
+      ..onHorizontalDragUpdate = _semanticsMapping.getHorizontalDragUpdateHandler(_findRecognizer) == null
+          ? null : _handleSemanticHorizontal
+      ..onVerticalDragUpdate = _semanticsMapping.getVerticalDragUpdateHandler(_findRecognizer) == null
+          ? null : _handleSemanticVertical;
   }
-
-  void _handleSemanticsLongPress() {
-    final LongPressGestureRecognizer recognizer = _recognizers[LongPressGestureRecognizer];
-    assert(recognizer != null);
-    if (recognizer.onLongPressStart != null)
-      recognizer.onLongPressStart(const LongPressStartDetails());
-    if (recognizer.onLongPress != null)
-      recognizer.onLongPress();
-    if (recognizer.onLongPressEnd != null)
-      recognizer.onLongPressEnd(const LongPressEndDetails());
-    if (recognizer.onLongPressUp != null)
-      recognizer.onLongPressUp();
+  GestureRecognizer _findRecognizer(Type type) => _recognizers[type];
+  GestureSemanticsMapping get _semanticsMapping => widget._semanticsMapping;
+  // The following methods handles semantic gestures, and are assigned to the
+  // render object [RenderSemanticsGestureHandler].
+  // They are assigned only when both semanticsMapping and the corresponding
+  // handler are non-null, therefore no null check of them is needed in the
+  // bodies. They are defined in a widget state because they need to be cached.
+  void _handleSemanticTap() {
+    _semanticsMapping.getTapHandler(_findRecognizer)();
```

Comment:
calling this twice (once in _assignSemantics and once here) is dangerous. It means that if the method is unreliable in what it returns, we might crash. For example, if it returns null the first time and non-null the second.

Reply:
Fixed. I've also added a test.